pkgpanda/build/__init__.py

```diff
@@ -456,61 +518,66 @@ def visit(pkg_tuple):
         assert pkg_tuple not in visited
         visited.add(pkg_tuple)
 
-        name = pkg_tuple[0]
-
-        # Ensure all dependencies are built
-        for require in sorted(packages[pkg_tuple].get("requires", list())):
+        # Ensure all dependencies are built. Sorted for stability
+        for require in sorted(package_store.packages[pkg_tuple]['requires']):
             require_tuple = expand_require(require)
             if require_tuple in built:
                 continue
             if require_tuple in visited:
-                raise BuildError("Circular dependency. Circular link {0} -> {1}".format(name, require_tuple))
+                raise BuildError("Circular dependency. Circular link {0} -> {1}".format(pkg_tuple, require_tuple))
 
             if PackageId.is_id(require_tuple[0]):
                 raise BuildError("Depending on a specific package id is not supported. Package {} "
-                                 "depends on {}".format(name, require_tuple))
+                                 "depends on {}".format(pkg_tuple, require_tuple))
 
-            if require_tuple not in packages:
-                raise BuildError("Package {0} require {1} not buildable from tree.".format(name, require_tuple))
+            if require_tuple not in package_store.packages:
+                raise BuildError("Package {0} require {1} not buildable from tree.".format(pkg_tuple, require_tuple))
 
             visit(require_tuple)
 
-        build_order.append(name)
+        build_order.append(pkg_tuple)
         built.add(pkg_tuple)
 
     # Can't compare none to string, so expand none -> "true" / "false", then put
     # the string in a field after "" if none, the string if not.
     def key_func(elem):
         return elem[0], elem[1] is None, elem[1] or ""
 
+    def visit_packages_in_tree(variant):
+        all_tuples = get_tree_package_tuples(package_store, variant)
+        for pkg_tuple in sorted(all_tuples, key=key_func):
+            if pkg_tuple in visited:
+                continue
+            visit(pkg_tuple)
+
     # Build everything if no variant is given
     if tree_variant is None:
```

Comment:
instead of doing this `if tree_variant is None` twice should switch to just making a ordered list to visit